### PR TITLE
8265900: Use pattern matching for instanceof at module jdk.compiler(part 2)

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This series of patches update the code of the module jdk.compiler by using the pattern matching for instanceof.

This patch would revise the following packages:
```
com.sun.tools.javac.code
com.sun.tools.javac.comp
com.sun.tools.javac.jvm
com.sun.tools.javac.parser
com.sun.tools.javac.processing
```

Thank you for taking the time to review.

--
Best Regards,
Guoxiong.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265900](https://bugs.openjdk.java.net/browse/JDK-8265900): Use pattern matching for instanceof at module jdk.compiler(part 2)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to 56970261ecbad153252cb454ebf830bf9bba0c9e

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3674/head:pull/3674` \
`$ git checkout pull/3674`

Update a local copy of the PR: \
`$ git checkout pull/3674` \
`$ git pull https://git.openjdk.java.net/jdk pull/3674/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3674`

View PR using the GUI difftool: \
`$ git pr show -t 3674`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3674.diff">https://git.openjdk.java.net/jdk/pull/3674.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Scope.java [hunk 1]
```diff
@@ -753,8 +753,8 @@ public void finalizeScope() {
         }
 
         protected Scope finalizeSingleScope(Scope impScope) {
-            if (impScope instanceof FilterImportScope && impScope.owner.kind == Kind.TYP &&
-                ((FilterImportScope) impScope).isStaticallyImported()) {
+            if (impScope instanceof FilterImportScope filterImportScope && impScope.owner.kind == Kind.TYP &&
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Scope.java:756]: break vertically on each `&&` ?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ArgumentAttr.java [hunk 2]
```diff
@@ -751,12 +751,9 @@ public int hashCode() {
 
         @Override
         public boolean equals(Object obj) {
-            if (obj instanceof UniquePos) {
-                UniquePos that = (UniquePos)obj;
-                return pos == that.pos && source == that.source;
-            } else {
-                return false;
-            }
+            return (obj instanceof UniquePos uniquePos)
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ArgumentAttr.java:754]: For another day - UniquePos can be a record :-)

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolConstant.java [hunk 3]
```diff
@@ -187,13 +187,9 @@ public int hashCode() {
...
             @Override
             public boolean equals(Object obj) {
-                if (obj instanceof BsmKey) {
-                    BsmKey other = (BsmKey)obj;
-                    return Objects.equals(bsmKey, other.bsmKey) &&
-                            Objects.equals(staticArgKeys, other.staticArgKeys);
-                } else {
-                    return false;
-                }
+                return (obj instanceof BsmKey key)
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolConstant.java:190]: BsmKey smells of records too...

**Review by mcimadamore — APPROVED** [review]
Looks good, one minor nit comment

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Scope.java (hunk 1 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Scope.java]: Fixed.

**Comment by lgxbslgx** [comment]
@mcimadamore I updated the code. Could I get your help to re-review and sponsor? Thanks a lot.

/integrate

**Pull request closed without merging**